File: workflow-orchestrator/src/main/java/com/omnibus/workflow/api/ApiErrorHandler.java
package com.omnibus.workflow.api;

import jakarta.validation.ConstraintViolationException;
import java.time.Instant;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.server.ServerWebInputException;

@RestControllerAdvice
public class ApiErrorHandler {

    @ExceptionHandler({
        IllegalArgumentException.class,
        ConstraintViolationException.class,
        ServerWebInputException.class
    })
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public Map<String, Object> handleBadRequest(Exception exception) {
        return Map.of(
            "timestamp", Instant.now().toString(),
            "message", exception.getMessage()
        );
    }

    @ExceptionHandler(IllegalStateException.class)
    @ResponseStatus(HttpStatus.BAD_GATEWAY)
    public Map<String, Object> handleIllegalState(IllegalStateException exception) {
        return Map.of(
            "timestamp", Instant.now().toString(),
            "message", exception.getMessage()
        );
    }
}

File: services/workflow-orchestrator/src/main/kotlin/pl/writeonly/omnibus/workflow/api/ApiErrorHandler.kt
package pl.writeonly.omnibus.workflow.api

import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.server.ServerWebInputException
import java.time.Instant

@RestControllerAdvice
class ApiErrorHandler {
    @ExceptionHandler(
        value = [
            IllegalArgumentException::class,
            ConstraintViolationException::class,
            ServerWebInputException::class,
        ],
    )
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleBadRequest(exception: Exception): Map<String, Any?> =
        mapOf(
            "timestamp" to Instant.now().toString(),
            "message" to exception.message,
        )

    @ExceptionHandler(IllegalStateException::class)
    @ResponseStatus(HttpStatus.BAD_GATEWAY)
    fun handleIllegalState(exception: IllegalStateException): Map<String, Any?> =
        mapOf(
            "timestamp" to Instant.now().toString(),
            "message" to exception.message,
        )
}

